crates/bevy_ecs/src/event.rs

```diff
@@ -141,66 +141,120 @@ impl<T> Default for Events<T> {
+            last_event_count: 0,
+            _marker: PhantomData,
+        }
     }
-}
-
-pub struct ManualEventReader<T> {
-    last_event_count: usize,
-    _marker: PhantomData<T>,
-}
 
-impl<T> Default for ManualEventReader<T> {
-    fn default() -> Self {
+    /// Gets a new [ManualEventReader]. This will ignore all events already in the event buffers. It
+    /// will read all future events.
+    pub fn get_reader_current(&self) -> ManualEventReader<T> {
         ManualEventReader {
-            last_event_count: 0,
-            _marker: Default::default(),
+            last_event_count: self.event_count,
+            _marker: PhantomData,
         }
     }
-}
 
-impl<T> ManualEventReader<T> {
-    /// See [`EventReader::iter`]
-    pub fn iter<'a>(&mut self, events: &'a Events<T>) -> impl DoubleEndedIterator<Item = &'a T> {
-        internal_event_reader(&mut self.last_event_count, events).map(|(e, _)| e)
+    /// Swaps the event buffers and clears the oldest event buffer. In general, this should be
+    /// called once per frame/update.
+    pub fn update(&mut self) {
+        match self.state {
+            State::A => {
+                self.events_b = Vec::new();
+                self.state = State::B;
+                self.b_start_event_count = self.event_count;
+            }
+            State::B => {
+                self.events_a = Vec::new();
+                self.state = State::A;
+                self.a_start_event_count = self.event_count;
+            }
+        }
     }
 
-    /// See [`EventReader::iter_with_id`]
-    pub fn iter_with_id<'a>(
-        &mut self,
-        events: &'a Events<T>,
-    ) -> impl DoubleEndedIterator<Item = (&'a T, EventId<T>)> {
-        internal_event_reader(&mut self.last_event_count, events)
+    /// A system that calls [Events::update] once per frame.
+    pub fn update_system(mut events: ResMut<Self>) {
+        events.update();
+    }
+
+    /// Removes all events.
+    pub fn clear(&mut self) {
+        self.events_a.clear();
+        self.events_b.clear();
+    }
+
+    /// Creates a draining iterator that removes all events.
+    pub fn drain(&mut self) -> impl DoubleEndedIterator<Item = T> + '_ {
+        self.drain_with_id().map(|(e, _)| e)
+    }
+
+    /// Creates a draining iterator that returns both events and their ids
+    pub fn drain_with_id(&mut self) -> impl DoubleEndedIterator<Item = (T, EventId<T>)> + '_ {
+        let event_instances = match self.state {
+            BufferState::A => self.events_b.drain(..).chain(self.events_a.drain(..)),
+            BufferState::B => self.events_a.drain(..).chain(self.events_b.drain(..)),
+        };
+
+        event_instances.map(|ei| {
+            trace!("Events::drain_with_id -> {}", ei.event_id);
+            (ei.event, ei.event_id)
+        })
+    }
+
+    pub fn extend<I>(&mut self, events: I)
+    where
+        I: Iterator<Item = T>,
+    {
+        for event in events {
+            self.send(event);
+        }
```

Comment:
It might be worth having `extend` internally call `self.events_X.extend(..)` as that'll end up having better performance since `Vec::extend` will outperform `Vec::push` in a loop. 

then have `Self::send` call 
```rust
self.extend(std::iter::once(event));
```

Reply:
I think that's a good change. Probably makes sense as a separate PR though?